### PR TITLE
[skip ci] [1st Pass] Initial draft for tether

### PR DESCRIPTION
**Pull request by caglar10ur**
This PR contains an initial design document as well as a -race free (according to golang race detector) - working prototype that implements the session handler and reaper

### CODE REVIEW

**Comment by caglar10ur** [comment]
to run the mocked tests

```
[conur@vir:/opt/go/src/github.com/vmware/vic(tether-ng)] cd lib/tether-ng/
/opt/go/src/github.com/vmware/vic/lib/tether-ng
[conur@vir:/opt/go/src/github.com/vmware/vic/lib/tether-ng] go generate && go test -v -race ./tests
=== RUN   TestRegister
--- PASS: TestRegister (0.00s)
=== RUN   TestRegisterMock
--- PASS: TestRegisterMock (0.00s)
=== RUN   TestRegisterFailure
--- PASS: TestRegisterFailure (0.00s)
=== RUN   TestConfigure
--- PASS: TestConfigure (0.00s)
PASS
ok      github.com/vmware/vic/lib/tether-ng/tests       1.029s
[conur@vir:/opt/go/src/github.com/vmware/vic/lib/tether-ng]
```

to interact with it (this will run bunch of commands in non-pty mode and will start bash in pty mode)

```
[vagrant@devbox:/opt/go/src/github.com/vmware/vic(tether-ng)] cd cmd/tether-ng/
/opt/go/src/github.com/vmware/vic/cmd/tether-ng
[vagrant@devbox:/opt/go/src/github.com/vmware/vic/cmd/tether-ng] go build -race && ./tether-ng && rm -f tether-ng
52f03eda-31e0-406d-bea0-750939a7624f
303ee1bf-a1de-43d2-9a23-c58f78d4fd03
ls is running
uname is running
ps is running
bash is running
Non-PseudoTermina ls
Runblock
WAITING
PseudoTerminal bash
Non-PseudoTermina uname
Releasing
Non-PseudoTermina ps
Really releasing (chan<- chan struct {})(0xc420012120)
Sleeping some before unblocking
stderr returned: %!s(<nil>)
...
```

**Comment by emlin** [comment]
Caglar, Could you please also add upgrade into the document? There have changes in the configuration part, and suppose will also change the interactive workflow between portlayer and tether.

**Comment by caglar10ur** [comment]
@emlin will add a section for migrating the configuration, the workflow between portlayer and tether will stay same - at least for a while.

File: lib/tether-ng/tether.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+type Plugin interface {
+	Configure(ctx context.Context, config *types.ExecutorConfig) error
+
+	Start(ctx context.Context) error
+	Stop(ctx context.Context) error
+
+	UUID(ctx context.Context) uuid.UUID
+}
+
+type Tetherer interface {
```
Comment (hmahmood) [inline lib/tether-ng/tether.go]: More like a PluginRegistrar?

Comment (caglar10ur) [inline lib/tether-ng/tether.go]: Yep, great suggestion, changing it

File: lib/tether-ng/toolbox/plugin.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func NewToolbox(ctx context.Context) *Toolbox {
+	return &Toolbox{
+		uuid: uuid.New(),
+		ctx:  ctx,
+	}
+}
+
+func (t *Toolbox) Configure(ctx context.Context, config *types.ExecutorConfig) error {
```
Comment (chengwang86) [inline lib/tether-ng/toolbox/plugin.go]: Do you want to replace the original Reload() function with this function? Or is this a totally different function?

Comment (caglar10ur) [inline lib/tether-ng/toolbox/plugin.go]: Please don't relay on anything on this PR as scrum priorities take over and I haven't touched this in this sprint and it is not clear when I'll be able return back to this

File: lib/tether-ng/process/plugin.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+func (p *Process) StartSession(ctx context.Context, i *types.Session) error {
+	in := make(chan *types.Session, 1)
+
+	// send session over and receive Done channel
+	in <- i
+	if i.Tty {
+		fmt.Printf("PseudoTerminal %s\n", i.ID)
+		i.Done = p.PseudoTerminal(ctx, in)
+	} else {
+		fmt.Printf("Non-PseudoTermina %s\n", i.ID)
```
Comment (sgairo) [inline lib/tether-ng/process/plugin.go]: "Non-PseudoTerminal "

File: lib/tether-ng/types/types.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
```
Comment (sgairo) [inline lib/tether-ng/types/types.go]: 2017

**Pull request closed without merging**